DittoHeartbeat/src/main/java/live.ditto.dittoheartbeat/Heartbeat.kt

```diff
@@ -0,0 +1,141 @@
+package live.ditto.dittoheartbeat
+
+import android.os.Build
+import androidx.annotation.RequiresApi
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.launch
+import live.ditto.Ditto
+import live.ditto.DittoConnectionType
+import live.ditto.DittoPeer
+import live.ditto.DittoSyncSubscription
+import org.joda.time.DateTime
+import java.math.BigInteger
+import java.security.MessageDigest
+import java.util.concurrent.atomic.AtomicBoolean
+
+data class HeartbeatConfig(
+    val id: Map<String, String>,
+    val interval: Long,
+    val collectionName: String,
+    val metaData: Map<String, Any>? = null
+)
+
+data class HeartbeatInfo(
+    val id: Map<String, String>,
+    val lastUpdated: String,
+    val presence: Presence?,
+)
+
+data class Presence(
+    val remotePeersCount: Int,
+    val peers: List<DittoPeer>,
+)
+
+var presence: Presence? = null
+var heartbeatSubscription: DittoSyncSubscription? = null
+
+@RequiresApi(Build.VERSION_CODES.O)
+fun startHeartbeat(ditto: Ditto, config: HeartbeatConfig): Flow<HeartbeatInfo> = flow {
+    val cancelable = AtomicBoolean(false)
+
+    if (heartbeatSubscription == null) {
+        heartbeatSubscription = ditto.sync.registerSubscription("SELECT * FROM ${config.collectionName}")
+    }
+
+    try {
+        while (!cancelable.get()) {
+            delay(config.interval)
+            val timestamp = DateTime().toISOString()
+            val info = HeartbeatInfo(
+                id = createCompositeId(config.id, ditto),
+                lastUpdated = timestamp,
+                presence = observePeers(ditto),
+            )
+            addToCollection(info, config, ditto)
+            emit(info)
+        }
+    } finally {
+        heartbeatSubscription!!.close()
+    }
+}
+
+fun createCompositeId(configId: Map<String, String>, ditto: Ditto): Map<String, String> {
+    val compositeId: MutableMap<String, String> = configId.toMutableMap()
+    val presenceGraph = ditto.presence.graph
+    compositeId["pk"] = byteArrayToHash(presenceGraph.localPeer.peerKey)
+    return compositeId
+}
+
+fun byteArrayToHash(byteArray: ByteArray): String {
+    val md5 = MessageDigest.getInstance("md5")
+
+    return BigInteger(1, md5.digest(byteArray))
+        .toString(16)
+        .padStart(32, '0')
+}
```

Comment:
I think we still need to change the logic here to be a base64 encoding of the bytes (URL safe) with the string "pk" stuck on the front.